L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+
+There is a lot of material on the Internet that covers Reactive Streams from many different angles. [ReactiveX](http://reactivex.io/) has lots of good documentation and [Marble diagrams](http://rxmarbles.com/) are a wonderful way to describe what Rx operators do.
+
+
+## Implementation
+
+A prototype of what is suggested in this RFC can be found [here](https://github.com/per-gron/shuriken/tree/master/src/rs-grpc). It implements the most important features, but it is still quite incomplete.
+
+Implementation of this RFC would involve:
+
+* Prepare the [rs](https://github.com/per-gron/shuriken/tree/master/src/rs) library so that it can be accepted as gRPC code.
+* Improve the prototype library so that it can be used for experimental purposes (fix memory leaks etc).
+* Write a protobuf code generator to be able to remove the most expensive C++ templates from the prototype and to make the API nicer.
+* Implement missing features in the prototype: Finish cancellation support, add support for timeouts, metadata etc.
+* Polish API and test.
+* Evangelize.
+
+The author could do much of this work, but it would be awesome to get help.
+
+
+## Open issues
+
+* The RPC handler methods might need to take one or more additional parameters to handle gRPC features like metadata.
+* This proposal glosses over many details; the API needs to be fleshed out more.
+* There are certain details in the current gRPC C++ API that seem to be missing: For example, the author could not find how get a signal in the server when a unary RPCs is cancelled.
+* The rs library is a fairly minimal implementation of Reactive Streams. One thing that many Rx libraries have that rs does not is a concept for a Publisher that emits only one value. This would increase the complexity of the library a bit, but it could be worth it for the additional type safety and performance benefits it would bring.
+* [rs](https://github.com/per-gron/shuriken/tree/master/src/rs) breaks Google's code style in the following ways. I am unsure about how exactly to fix this.
+  * **It uses C++14:** This is mostly for ease of implementation; it should be possible to backport to C++11.
+  * **It uses rvalue references:** These could mostly be changed to by-value instead, but I have a hard time justifying the performance penalty.
+  * **It uses `std::exception_ptr`:** Reactive Streams has built in error handling that is based on having a generic type for errors. `std::exception_ptr` is C++'s type for that. rs could use something else, but the chances of being able to convince the Reactive Streams people that rs should become the official C++ version of Reactive Streams is likely low without `std::exception_ptr`. Please note that rs never throws, it uses `std::exception_ptr` only as a container type.
```

Comment:
grpc also has a generic container for error handling (grpc::Status).

It's hard to justify introducing a wrinkle that forces conversion.

Reply:
I understand that. I am not opposed to changing this part of rs.

Perhaps it would be possible to change it so that the error type is configurable via template or something. But if it becomes too complicated I think it could be okay to tie it to gRPC's types.

Reply:
I have added a bullet point under the "Implementation" section to change rs so that `grpc::Status` can be used as the error type.

With this I consider this issue resolved, let me know if you disagree.